File: 64.minimum-path-sum.js
/*
 * @lc app=leetcode id=64 lang=javascript
 *
 * [64] Minimum Path Sum
 */

// @lc code=start
/**
 * @param {number[][]} grid
 * @return {number}
 */
var minPathSum = function (grid) {
  const len = grid.length;
  const dp = Array.from({ length: len }, () => []);
  dp[0][0] = grid[0][0];
  for (let i = 0; i < len; ++i) {
    for (let j = 0; j < len; ++j) {
      if (!i && !j) continue;
      dp[i][j] =
        grid[i][j] +
        Math.min(dp[Math.max(i - 1, 0)][j], dp[i][Math.max(j - 1, 0)]);
    }
  }
  return dp[len - 1][len - 1];
};
// @lc code=end

File: solutions/64.minimum-path-sum.rb
#
# @lc app=leetcode id=64 lang=ruby
#
# [64] Minimum Path Sum
#

# @lc code=start
# @param {Integer[][]} grid
# @return {Integer}
def min_path_sum(grid)
  map = {}

  dfs = lambda do |x, y|
      return map[[x,y]] if map.key? [x,y]
      return grid[0][0] if x==0 && y==0
      return grid[x][y] + dfs.call(x, y-1) if x==0
      return grid[x][y] + dfs.call(x-1, y) if y==0
      map[[x,y]] ||= grid[x][y] + [dfs.call(x, y-1), dfs.call(x-1, y)].min
  end

  m = grid.length
  n = grid[0].length
  dfs.call(m-1, n-1)
end
# @lc code=end

